### PR TITLE
Feat: Add support for `dio`

### LINKED ISSUE #673: Support for DIO HTTP library
[dio](https://pub.dev/packages/dio) is a popular HTTP library and I would like to request support for it, similar to the already existing support for the [SentryHttpClient](https://github.com/getsentry/sentry-dart/blob/main/dart/lib/src/http_client/sentry_http_client.dart).

They have support for so-called [interceptors](https://pub.dev/packages/dio#interceptors), which should make it straight forward and also pretty similar to the `SentryHttpClient`.

### PR DESCRIPTION
**Pull request by ueman**
## :scroll: Description
See https://github.com/getsentry/sentry-dart/issues/673
Bildschirmfoto 2022-01-11 um 20 02 56

## :bulb: Motivation and Context
See https://github.com/getsentry/sentry-dart/issues/673

Closes #673

## :green_heart: How did you test it?
New tests & CI pipeline

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] I updated the docs if needed
- [x] All tests passing
- [x] No breaking changes

## :crystal_ball: Next steps
Publish to pub.dev

### CODE REVIEW

File: dart/lib/sentry.dart [hunk 1]
```diff
@@ -22,6 +22,7 @@ export 'src/transport/transport.dart';
 export 'src/integration.dart';
 export 'src/event_processor.dart';
 export 'src/http_client/sentry_http_client.dart';
+export 'src/http_client/sentry_http_client_error.dart';
```
Comment (ueman) [inline dart/lib/sentry.dart:25]: This is exported in order to be used by `sentry_dio`s `FailedRequestClientAdapter`.

File: dio/lib/sentry_dio.dart [hunk 2]
```diff
@@ -0,0 +1,3 @@
+library sentry_dio;
+
+export 'src/sentry_client_adapter.dart';
```
Comment (ueman) [inline dio/lib/sentry_dio.dart]: Do we want to export each individual client adapter? Probably not, right?

File: dio/pubspec.yaml [hunk 3]
```diff
@@ -0,0 +1,22 @@
+name: sentry_dio
+description: An integration which adds support for performance tracing for the Dio package.
+version: 6.3.0
+homepage: https://docs.sentry.io/platforms/dart/
+repository: https://github.com/getsentry/sentry-dart
+issue_tracker: https://github.com/getsentry/sentry-dart/issues
```
Comment (ueman) [inline dio/pubspec.yaml:6]: The other packages don't have the `issue_tracker` field, so it should probably aligned?

File: dio/pubspec.yaml [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+environment:
+  sdk: '>=2.12.0 <3.0.0'
+
+
+dependencies:
+  dio: ^4.0.0
+  sentry: 
+    path: ../dart
```
Comment (ueman) [inline dio/pubspec.yaml]: I had to change the pure dart sdk, so I need to use a path dependency. Obviously, this must be changed before the release.

File: dio/lib/src/adapter/dio_adapter.dart [hunk 5]
```diff
@@ -0,0 +1,9 @@
+// coverage:ignore-file
+// ignore_for_file: public_member_api_docs
+
+import 'package:dio/dio.dart';
+
+import '_io_adapter.dart' if (dart.library.html) '_browser_adapter.dart'
+    as adapter;
+
+HttpClientAdapter createAdapter() => adapter.createAdapter();
```
Comment (ueman) [inline dio/lib/src/adapter/dio_adapter.dart]: Needed because https://github.com/flutterchina/dio/issues/1362

File: dio/pubspec.yaml (hunk 4 above)
Comment (kuhnroyal) [reply dio/pubspec.yaml]: This prevents usage as git dependency, better define the version here and add the path to dependency_overrides.

File: dio/lib/src/breadcrumb_client_adapter.dart [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+/// which records requests as breadcrumbs.
+///
+/// Remarks:
+/// If this client is used as a wrapper, a call to close also closes the
+/// given client.
+class BreadcrumbClientAdapter extends HttpClientAdapter {
+  // ignore: public_member_api_docs
+  BreadcrumbClientAdapter({HttpClientAdapter? client, Hub? hub})
+      : _hub = hub ?? HubAdapter(),
+        _client = client ?? createAdapter();
```
Comment (kuhnroyal) [inline dio/lib/src/breadcrumb_client_adapter.dart]: What about not using a default adapter here and making the client required. That would avoid all the dio export mess.

Comment (ueman) [reply dio/lib/src/breadcrumb_client_adapter.dart]: That would just move the burden of creating this code to the user :(

Or could the iser do something like:
```dart
final dio = Dio();
dio.httpClientAdapter = SentryHttpClientAdapter(client: dio.httpClientAdapter);
```

I guess that would be acceptable.

File: dio/lib/src/breadcrumb_client_adapter.dart [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    RequestOptions options,
+    Stream<Uint8List>? requestStream,
+    Future? cancelFuture,
+  ) async {
+    // See https://develop.sentry.dev/sdk/event-payloads/breadcrumbs/
+
+    var requestHadException = false;
+    int? statusCode;
+    String? reason;
+    int? responseBodySize;
```
Comment (ueman) [inline dio/lib/src/breadcrumb_client_adapter.dart:34]: This is currently unused. @kuhnroyal is it somehow possible to get the response size?

File: dio/lib/src/failed_request_client_adapter.dart [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+    // As far as I can tell there's no way to get the uri without the query part
+    // so we replace it with an empty string.
+    final urlWithoutQuery = options.uri.replace(query: '').toString();
+
+    final query = options.uri.query.isEmpty ? null : options.uri.query;
+
+    final headers = options.headers
+        .map((key, dynamic value) => MapEntry(key, value?.toString() ?? ''));
+
+    final sentryRequest = SentryRequest(
```
Comment (ueman) [inline dio/lib/src/failed_request_client_adapter.dart:131]: Ideally we would record the content and content size but as far as I can tell, that's not possible. Is that right, @kuhnroyal ?

File: dio/pubspec.yaml (hunk 4 above)
Comment (ueman) [reply dio/pubspec.yaml]: Done

File: dio/lib/src/breadcrumb_client_adapter.dart (hunk 6 above)
Comment (ueman) [reply dio/lib/src/breadcrumb_client_adapter.dart]: Changed it

File: dio/lib/sentry_dio.dart (hunk 2 above)
Comment (ueman) [reply dio/lib/sentry_dio.dart]: Nope, decided to go with another solution

File: dio/lib/src/sentry_transformer.dart [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+      : _hub = hub ?? HubAdapter(),
+        _transformer = transformer;
+
+  final Transformer _transformer;
+  final Hub _hub;
+
+  @override
+  Future<String> transformRequest(RequestOptions options) async {
+    final span = _hub.getSpan()?.startChild(
+          'transform-request',
```
Comment (ueman) [inline dio/lib/src/sentry_transformer.dart]: According to https://develop.sentry.dev/sdk/performance/span-operations/ this should probably be `serialize`?

File: dio/example/example.dart [hunk 10]
```diff
@@ -0,0 +1,17 @@
+import 'package:dio/dio.dart';
+import 'package:sentry/sentry.dart';
+import 'package:sentry_dio/sentry_dio.dart';
+
+Future<void> main() async {
```
Comment (marandaneto) [inline dio/example/example.dart:5]: can we make a pub project like https://github.com/getsentry/sentry-dart/tree/main/dart/example?
so we can compile and run easily on CI as well

File: dio/lib/src/sentry_dio_extension.dart [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+/// Extension to add performance tracing for [Dio]
+extension SentryDioExtension on Dio {
+  /// Adds support for automatic spans for http requests,
+  /// as well as request and response transformations.
+  /// This must be the last initialization step of the [Dio] setup, otherwise
+  /// your configuration of Dio might overwrite the Sentry configuration.
+  void addSentry({
+    bool recordBreadcrumbs = true,
+    bool networkTracing = true,
+    MaxRequestBodySize maxRequestBodySize = MaxRequestBodySize.never,
```
Comment (marandaneto) [inline dio/lib/src/sentry_dio_extension.dart:15]: isn't `small` the default value?

File: dio/example/example.dart (hunk 10 above)
Comment (ueman) [reply dio/example/example.dart:5]: Sure. I've also included it in the Flutter example app, maybe that's already enough?

File: dio/lib/src/sentry_client_adapter.dart [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+/// Set [networkTracing] to `true` to enable it.
+///
+/// Remarks: If this client is used as a wrapper, a call to close also closes
+/// the given client.
+///
+/// Remarks:
+/// HTTP traffic can contain PII (personal identifiable information).
+/// Read more on data scrubbing [here](https://docs.sentry.io/product/data-management-settings/advanced-datascrubbing/).
+/// ```
+class SentryHttpClientAdapter extends HttpClientAdapter {
```
Comment (marandaneto) [inline dio/lib/src/sentry_client_adapter.dart]: for some reason HJ isn't highlighting the colors here

File: dio/lib/src/sentry_client_adapter.dart [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+/// through the [TracingClientAdapter] client, it's disabled by default.
+/// Set [networkTracing] to `true` to enable it.
+///
+/// Remarks: If this client is used as a wrapper, a call to close also closes
+/// the given client.
+///
+/// Remarks:
+/// HTTP traffic can contain PII (personal identifiable information).
+/// Read more on data scrubbing [here](https://docs.sentry.io/product/data-management-settings/advanced-datascrubbing/).
+/// ```
```
Comment (ueman) [inline dio/lib/src/sentry_client_adapter.dart]:
```suggestion
```

File: dio/lib/src/sentry_client_adapter.dart (hunk 12 above)
Comment (ueman) [reply dio/lib/src/sentry_client_adapter.dart]: Probably tripping over the backticks in the line above. I've removed them. I hope that should fix it.

Comment (marandaneto) [inline dio/lib/src/sentry_client_adapter.dart]: what's about `SentryDioClientAdapter`? does it make sense?

Comment (ueman) [reply dio/lib/src/sentry_client_adapter.dart]: Sounds good. I shouldn't forgot to rename the file when renaming it.

File: dio/lib/src/sentry_dio_extension.dart [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+      recordBreadcrumbs: recordBreadcrumbs,
+      networkTracing: networkTracing,
+      maxRequestBodySize: maxRequestBodySize,
+      failedRequestStatusCodes: failedRequestStatusCodes,
+      captureFailedRequests: captureFailedRequests,
+      sendDefaultPii: sendDefaultPii,
+    );
+
+    // intercept transformations
+    transformer = SentryTransformer(transformer: transformer);
```
Comment (marandaneto) [inline dio/lib/src/sentry_dio_extension.dart:32]: where does `transformer` come from? inherited from `Dio`?

Comment (ueman) [reply dio/lib/src/sentry_dio_extension.dart:32]: Yes. The transformer is responsible for (de)serialization of the http requests and responses.

File: dio/lib/src/sentry_transformer.dart [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+          'serialize',
+          description: 'Dio.transformRequest: ${options.method} ${options.uri}',
+        );
+    try {
+      final request = await _transformer.transformRequest(options);
+      await span?.finish(status: SpanStatus.ok());
+      return request;
+    } catch (_) {
+      await span?.finish(status: SpanStatus.internalError());
+      rethrow;
```
Comment (marandaneto) [inline dio/lib/src/sentry_transformer.dart]: I'd rather use the convention used by `TracingClient`, which is the try/catch/finally to set status, exception, rethrow and finish

File: dio/lib/src/sentry_dio_extension.dart (hunk 11 above)
Comment (ueman) [reply dio/lib/src/sentry_dio_extension.dart:15]: Not sure about that, but I guess not as I've copied it from the HTTP integration. I'll double check it though.

File: dio/lib/src/sentry_transformer.dart [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+        _transformer = transformer;
+
+  final Transformer _transformer;
+  final Hub _hub;
+
+  @override
+  Future<String> transformRequest(RequestOptions options) async {
+    final span = _hub.getSpan()?.startChild(
+          'serialize',
+          description: 'Dio.transformRequest: ${options.method} ${options.uri}',
```
Comment (marandaneto) [inline dio/lib/src/sentry_transformer.dart]: `serialize` already follows the convention https://develop.sentry.dev/sdk/performance/span-operations/#web-server
I'd rather remove the `Dio.transformRequest` prefix from the description, wdyt?

File: dio/lib/src/sentry_transformer.dart [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+      return request;
+    } catch (_) {
+      await span?.finish(status: SpanStatus.internalError());
+      rethrow;
+    }
+  }
+
+  @override
+  // ignore: strict_raw_type
+  Future transformResponse(
```
Comment (marandaneto) [inline dio/lib/src/sentry_transformer.dart:36]: same comments as transformRequest

File: dio/lib/src/sentry_transformer.dart (hunk 16 above)
Comment (ueman) [reply dio/lib/src/sentry_transformer.dart]: I'm always quite unsure what to use as operation and description. So if you're saying dropping the prefix makes sense it's fine for me. I've included it, so that you're able to see where it comes from.

**Comment by marandaneto** [comment]
@ueman can we possibly get rid of the `ignore_for_file` and friends by fixing the issues if necessary? unless there's no reason to do so

**Comment by marandaneto** [comment]
@ueman that's awesome, thank you so much, left a few comments but I focused on structure and design only, I'll let @brustolin or @denrase with a more detailed review.
very happy about the direction of this PR, congrats.

**Comment by ueman** [comment]
> @ueman can we possibly get rid of the `ignore_for_file` and friends by fixing the issues if necessary? unless there's no reason to do so

[Unfortunately dio isn't configured as strict](https://github.com/flutterchina/dio/issues/1372) as this package, so I've included it in each file which interacts with dios API. Being stricter than dio can lead to runtime errors. If there are any better ways to handle this, I'm happy to hear about it.

**Comment by marandaneto** [comment]
@ueman mind attaching a screenshot of the Dio request using the sample? with the automatically created spans.

**Comment by ueman** [comment]
> @ueman mind attaching a screenshot of the Dio request using the sample? with the automatically created spans.

@marandaneto I'v attached it to the PR description at the top.

File: dio/example/example.dart (hunk 10 above)
Comment (marandaneto) [reply dio/example/example.dart:5]: mm I guess because of `paths-ignore`, we'd not have a failed CI when changes don't integrate well?
I still kinda prefer the pub project facility, but it's a personal opinion.

File: dio/lib/src/sentry_dio_extension.dart (hunk 11 above)
Comment (marandaneto) [reply dio/lib/src/sentry_dio_extension.dart:15]: its `never`, my bad

File: dio/lib/src/sentry_transformer.dart (hunk 16 above)
Comment (marandaneto) [reply dio/lib/src/sentry_transformer.dart]: both are raw Strings and there's a convention, but its not enforced anywhere, some UI features rely on the span operation, that's why the docs page, description is open, yet.
I'd remove the `Dio.transformRequest:` since the method/ URL and span operation already tells you where it comes from

**Review by brustolin — APPROVED** [review]
I really like this solution! 🚀

File: flutter/analysis_options.yaml [hunk 18]
```diff
@@ -13,8 +13,7 @@ analyzer:
     deprecated_member_use_from_same_package: warning
     # ignore sentry/path on pubspec as we change it on deployment
     invalid_dependency: ignore
-  exclude:
-    - example/**
```
Comment (marandaneto) [inline flutter/analysis_options.yaml:17]: Why is that removed?

File: flutter/analysis_options.yaml [hunk 19]
```diff
@@ -13,8 +13,7 @@ analyzer:
...
+    unnecessary_import: ignore
```
Comment (marandaneto) [inline flutter/analysis_options.yaml]: Did we not fix all the issues? This is not necessary anymore

Comment (ueman) [reply flutter/analysis_options.yaml]: You're probably right. This must be a merge issue.

File: flutter/analysis_options.yaml (hunk 18 above)
Comment (ueman) [reply flutter/analysis_options.yaml:17]: This allows the example to be analysed and linted. That's also why there are so much changes in the example, because I've fixed the warnings for the example.

File: dio/example/example.dart (hunk 10 above)
Comment (marandaneto) [reply dio/example/example.dart:5]: we can do that in a different PR, so we don't block it.

**Comment by marandaneto** [comment]
@ueman happy to merge and release after addressing https://github.com/getsentry/sentry-dart/pull/688#discussion_r784747005
thanks once more.

**Review by marandaneto — APPROVED** [review]

File: flutter/analysis_options.yaml (hunk 19 above)
Comment (ueman) [reply flutter/analysis_options.yaml]:
```suggestion
```

**Comment by marandaneto** [comment]
@ueman 

> warning • The include file 'package:pedantic/analysis_options.yaml' in '/home/runner/work/sentry-dart/sentry-dart/flutter/analysis_options.yaml' can't be found when analyzing '/home/runner/work/sentry-dart/sentry-dart/flutter/example' • analysis_options.yaml:1:10 • include_file_not_found
>    info • The import of 'package:flutter/rendering.dart' is unnecessary because all of the used elements are also provided by the import of 'package:flutter/material.dart' • example/lib/user_feedback_dialog.dart:2:8 • unnecessary_import
